### ISSUE TITLE
Invalid mac address returned if `ip` prints to stderr

### ISSUE THREAD

**Issue by cutwater** [projectcalico/libcalico#148]
Function `get_ns_veth_mac` [netns.py#L245](https://github.com/projectcalico/libcalico/blob/master/calico_containers/pycalico/netns.py#L245) returns invalid result if `ip netns ...` command printed message in stderr.

`ip netns` command can print `RTNETLINK answers: Invalid argument` to stderr and still finish with exitcode `0`, which causes next endpoint written to etcd:

`{"name": "cali76e50b0684a", "labels": {}, "state": "active", "ipv6_nets": [], "mac": "RTNETLINK answers: Invalid argument\n2a:1b:78:a8:63:a4", "ipv4_nets": ["10.233.67.164/32"], "profile_ids": ["calico-k8s-network"]}`

The root cause of this is that function [NamedNamespace.check_output](https://github.com/projectcalico/libcalico/blob/master/calico_containers/pycalico/netns.py#L345) redirects **stderr** to **stdout**.

Solution is to read output only from **stdout**.

**Comment by matthewdupre** [comment]
Assigning this a low priority on the assumption that the original cause (of ip netns erroring) can be worked around: I can adjust if it can't.  @cutwater did you find out why the command errored?

**Comment by cutwater** [comment]
@matthewdupre We're still investigating the root cause. This issue occurs when there is something wrong with linux namespaces (usually after OpenStack neutron creates `dhcp-` namespace). Still `ip netns` works correctly, returns the result as expected and exitcode is `0`, except it has such side-effect like writing message to _stderr_.

The problems begin when `NamedNamespace` reads stderr, considering it as a result of program execution. In my opinion this is incorrect behaviour, which affects data that is written to etcd.

**Comment by nelljerram** [comment]
@djosborne You added the `stderr=STDOUT` as part of:

```
65e86e67980e6efe7461b5ea6427067e2f45c332
Author:     Dan Osborne <daniel.osborne@metaswitch.com>
AuthorDate: Tue Oct 13 16:27:16 2015 -0700
Commit:     Dan Osborne <daniel.osborne@metaswitch.com>
CommitDate: Tue Oct 13 16:27:16 2015 -0700

Removed use of check_call in netns and improved logging

Replaced with check_output, which, when it throws an exception, includes a .output field which can be used to better log encountered errors
```

Can you remember why?

**Comment by ozdanborne** [comment]
@neiljerram That was back when I was trying to improve error logging and diags. I followed [the example on the python docs](https://docs.python.org/2/library/subprocess.html#subprocess.check_output):

> To also capture standard error in the result, use stderr=subprocess.STDOUT:
> 
> > > subprocess.check_output(
> > > ...     "ls non_existent_file; exit 0",
> > > ...     stderr=subprocess.STDOUT,
> > > ...     shell=True)
> > > 'ls: non_existent_file: No such file or directory\n'

When commands failed and printed why to stderr, libcalico could bubble up the reason why to the caller.

As @cutwater has shown, redirecting stderr is a bit overreaching. If we remove it, error messages may become a bit less descript, but we can just find another way to improve that.

**Comment by nelljerram** [comment]
Thanks @djosborne.

@cutwater Do you have any more information about the root cause when `ip netns exec ...` returns 0 but generates stuff on stderr?

**Comment by adidenko** [comment]
@neiljerram here's how you can reproduce it:

```
mkdir /var/run/netns/zzz
ip netns
RTNETLINK answers: Invalid argument
RTNETLINK answers: Invalid argument
zzz
```

**Comment by nelljerram** [comment]
@adidenko For me that doesn't give the symptoms that this PR is concerned with:

```
root@smaug:~# mkdir /var/run/netns
root@smaug:~# mkdir /var/run/netns/zzz
root@smaug:~# ip netns
RTNETLINK answers: Invalid argument
RTNETLINK answers: Invalid argument
zzz
root@smaug:~# ip netns exec zzz ip link
RTNETLINK answers: Invalid argument
setting the network namespace "zzz" failed: Invalid argument
root@smaug:~# echo $?
1
```

Were there some other details to your repro?

**Comment by adidenko** [comment]
That's exactly the case we faced. There were some leftovers in /var/run/netns directory on kubernetes worker nodes and this caused invalid "mac" parameter for newly created PODs on such nodes.

So you can do `mkdir /var/run/netns/zzz` on any kubernetes worker node and then try to spawn any POD on it.

**Comment by nelljerram** [comment]
(1) But why would anyone do that? (Instead of creating a namespace properly with `ip netns add`.)

(2) The PR discussion above says that part of the problem is that the exit code is 0, but my transcript just above shows that I get an exit code of 1.

I'm afraid I'm quite confused!

**Comment by adidenko** [comment]
1) for example, some POD/container may mount `/run/netns` (or `/run`) from host system and create some netns for its needs. Then after removal of this POD it will leave the netns bind on host system.

2) `ip netns` (i.e. `ip netns ls`) works with exit code 0

```
$ touch /var/run/netns/zzz
$ ip netns ; echo $?
RTNETLINK answers: Invalid argument
RTNETLINK answers: Invalid argument
zzz
0
```

So if you have a valid netns and a leftover in /run/netns dir, the valid netns works fine but still throws warnings:

```
$ ip netns add okns
$ touch /var/run/netns/zzz

$ ip netns ls ; echo $?
RTNETLINK answers: Invalid argument
RTNETLINK answers: Invalid argument
zzz
okns
0

$ ip netns exec okns ip a ; echo $?
RTNETLINK answers: Invalid argument
1: lo: <LOOPBACK> mtu 65536 qdisc noop state DOWN group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
0
```

We do not intend to use this broken `zzz` netns in Calico. It just causes those `RTNETLINK answers: Invalid argument` warnings on any `ip netns *` command execution which breaks networking for Kubernetes.

**Comment by nelljerram** [comment]
Thanks @adidenko, I see now, and agree that this is a strong argument for not having `stderr=STDOUT` in `ip netns` calls.  It basically means that our code can never reliably use the expected output of such a call.

**Comment by nelljerram** [comment]
Fixed by PR #149
